File: simpleagent.cpp
#include <iostream>
#include <string>
#include <vector>

class Environment; // Forward declaration

class Sensor
{
public:
    Sensor(Environment *environment) : environment(environment) {}

    std::vector<std::string> getPercept();

private:
    Environment *environment;
};

class Actuator
{
public:
    Actuator(Environment *environment) : environment(environment) {}

    void performAction(std::pair<std::string, std::string> action);

private:
    Environment *environment;
};

class Environment
{
public:
    Environment()
    {
        locationA = "Dirty";
        locationB = "Dirty";
    }

    std::vector<std::string> getPercept()
    {
        return {locationA, locationB};
    }

    void cleanLocation(std::string location)
    {
        if (location == "A")
        {
            locationA = "Clean";
        }
        else if (location == "B")
        {
            locationB = "Clean";
        }
    }

private:
    std::string locationA;
    std::string locationB;
};

std::vector<std::string> Sensor::getPercept()
{
    return environment->getPercept();
}

void Actuator::performAction(std::pair<std::string, std::string> action)
{
    std::string location = action.first;
    std::string actionType = action.second;

    if (actionType == "Suck")
    {
        environment->cleanLocation(location);
    }
}

class Agent
{
public:
    Agent(Environment *environment) : sensor(environment), actuator(environment) {}

    std::vector<std::string> perceive()
    {
        return sensor.getPercept();
    }

    void act()
    {
        std::vector<std::string> percept = perceive();
        std::string locationA = percept[0];
        std::string locationB = percept[1];

        std::pair<std::string, std::string> action;

        if (locationA == "Dirty")
        {
            action = std::make_pair("A", "Suck");
        }
        else if (locationB == "Dirty")
        {
            action = std::make_pair("B", "Suck");
        }
        else
        {
            action = (locationA == "Clean") ? std::make_pair("A", "Move") : std::make_pair("B", "Move");
        }

        actuator.performAction(action);
    }

private:
    Sensor sensor;
    Actuator actuator;
};

int main()
{
    // Create an environment
    Environment environment;

    // Create a vacuum cleaner agent with the environment
    Agent vacuumAgent(&environment);

    // Agent performs an action
    vacuumAgent.act();

    // Agent perceives the updated state
    std::vector<std::string> percept = vacuumAgent.perceive();
    std::cout << "Percept: " << percept[0] << ", " << percept[1] << std::endl;

    return 0;
}

File: simple_agent.py
class Sensor:
    def __init__(self, environment):
        self.environment = environment

    def get_percept(self):
        # Get the current percept from the environment
        return self.environment.get_percept()


class Environment:
    def __init__(self):
        self.location_a = "Dirty"
        self.location_b = "Dirty"

    def get_percept(self):
        # Return the current percept, indicating whether the current location is dirty or not
        return self.location_a, self.location_b

    def clean_location(self, location):
        # Clean the specified location
        if location == "A":
            self.location_a = "Clean"
        elif location == "B":
            self.location_b = "Clean"


class Actuator:
    def __init__(self, environment):
        self.environment = environment

    def perform_action(self, action):
        # Perform the given action on the environment
        location, action_type = action
        if action_type == "Suck":
            self.environment.clean_location(location)


class Agent:
    def __init__(self, environment):
        self.sensor = Sensor(environment)
        self.actuator = Actuator(environment)

    def perceive(self):
        # Get the current percept from the sensor
        return self.sensor.get_percept()

    def act(self):
        # Decide and perform an action based on the percept
        location_a, location_b = self.perceive()

        if location_a == "Dirty":
            action = ("A", "Suck")
        elif location_b == "Dirty":
            action = ("B", "Suck")
        else:
            # If both locations are clean, move randomly to another location
            action = ("A" if location_a == "Clean" else "B", "Move")

        self.actuator.perform_action(action)


if __name__ == '__main__':
    # Create an environment
    environment = Environment()

    # Create a vacuum cleaner agent with the environment
    vacuum_agent = Agent(environment)

    # Agent performs an action
    vacuum_agent.act()

    # Agent perceives the updated state
    percept = vacuum_agent.perceive()
    print("Percept:", percept)
